shell/browser/api/electron_api_view.cc

```diff
@@ -4,34 +4,357 @@
+  static bool FromV8(v8::Isolate* isolate,
+                     v8::Local<v8::Value> val,
+                     views::FlexAllocationOrder* out) {
+    std::string orientation = base::ToLowerASCII(gin::V8ToString(isolate, val));
+    if (orientation == "normal") {
+      *out = views::FlexAllocationOrder::kNormal;
+    } else if (orientation == "reverse") {
+      *out = views::FlexAllocationOrder::kReverse;
+    } else {
+      return false;
+    }
+    return true;
+  }
+};
+
+template <>
+struct Converter<views::SizeBound> {
+  static v8::Local<v8::Value> ToV8(v8::Isolate* isolate,
+                                   const views::SizeBound& in) {
+    if (in.is_bounded())
+      return v8::Integer::New(isolate, in.value());
+    return v8::Number::New(isolate, std::numeric_limits<double>::infinity());
+  }
+};
+
+template <>
+struct Converter<views::SizeBounds> {
+  static v8::Local<v8::Value> ToV8(v8::Isolate* isolate,
+                                   const views::SizeBounds& in) {
+    return gin::DataObjectBuilder(isolate)
+        .Set("width", in.width())
+        .Set("height", in.height())
+        .Build();
+  }
+};
+}  // namespace gin
 
 namespace electron::api {
 
+using LayoutCallback = base::RepeatingCallback<views::ProposedLayout(
+    const views::SizeBounds& size_bounds)>;
+
+class JSLayoutManager : public views::LayoutManagerBase {
+ public:
+  explicit JSLayoutManager(LayoutCallback layout_callback)
+      : layout_callback_(std::move(layout_callback)) {}
+  ~JSLayoutManager() override {}
+
+  views::ProposedLayout CalculateProposedLayout(
+      const views::SizeBounds& size_bounds) const override {
+    v8::Isolate* isolate = JavascriptEnvironment::GetIsolate();
+    v8::HandleScope handle_scope(isolate);
+    return layout_callback_.Run(size_bounds);
+  }
+
+ private:
+  LayoutCallback layout_callback_;
+};
+
 View::View(views::View* view) : view_(view) {
   view_->set_owned_by_client();
+  view_->AddObserver(this);
 }
 
 View::View() : View(new views::View()) {}
 
 View::~View() {
+  if (!view_)
+    return;
+  view_->RemoveObserver(this);
   if (delete_view_)
     delete view_;
 }
 
+void View::AddChildViewAt(gin::Handle<View> child,
+                          absl::optional<size_t> maybe_index) {
+  // TODO(nornagon): !view_ is only for supporting the weird case of
+  // WebContentsView's view being deleted when the underlying WebContents is
+  // destroyed (on non-Mac). We should fix that so that WebContentsView always
+  // has a View, possibly a wrapper view around the underlying platform View.
+  if (!view_)
+    return;
+  size_t index =
+      std::min(child_views_.size(), maybe_index.value_or(child_views_.size()));
+  child_views_.emplace(child_views_.begin() + index,     // index
+                       isolate(), child->GetWrapper());  // v8::Global(args...)
+#if BUILDFLAG(IS_MAC)
+  // Disable the implicit CALayer animations that happen by default when adding
+  // or removing sublayers.
```

Comment:
Reading this reminded me of Electron's private API patches for the MAS build. Is there potential for any of the View changes to affect that build?

Reply:
Hm... is this just because a CALayer is involved? I think the MAS-related stuff for CALayer is stuff to do with private APIs for cross-process CALayer things, which this doesn't affect as everything here is in-process and using public APIs.